tensorboard/components/tensor_widget/types.ts

```diff
@@ -0,0 +1,208 @@
+ * widget, this interface should have the following concrete value:
+ *
+ * ```
+ * {
+ *   slicingDimsAndIndices: [{
+ *     dim: 0,
+ *     index: 0
+ *   }, {
+ *     dim: 3,
+ *     index: 2
+ *   }],
+ *   viewingDims: [1, 2],
+ *   verticalRange: [0, 4],
+ *   horizontalRange: [0, 3]
+ * }
+ * ```
+ */
+export interface TensorViewSlicingSpec {
+  /**
+   * Which dimensions of the tensor are sliced down to a slice of 1.
+   *
+   * - The `dim` field is the 0-based dimension index.
+   * - The `index` is the 0-based index for the selected slice.
+   */
+  slicingDimsAndIndices: Array<{dim: number, index: number}>;
+
+  /**
+   * Which dimensions are used for viewing (i.e., rendered in the
+   * tensor widget, as a table, by default.)
+   *
+   * Possible lengths of this array field:
+   * - 0 for scalar tensors.
+   * - 1 for 1D tensors.
+   * - 2 for 2D+ tensors.
+   */
+  viewingDims: number[];
+
+  /**
+   * The indices from the first viewing dimension, which are shown as rows.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  verticalRange: [number, number];
+
+  /**
+   * The indices from the second viewing dimension, which are shown as columns.
+   *
+   * The two numbers are beginning index (inclusive) and ending index
+   * (exclusive).
+   */
+  horizontalRange?: [number, number];
+
+  /**
+   * Optional dimension for depth.
+   *
+   * This supports visualization that requires a depth dimension, e.g.,
+   * color channels.
+   */
+  depthDim?: number;
+}
+
+/** Options used during the creation of a single-tensor tensor widget. */
+export interface TensorWidgetOptions {
+  /** Name of the tensor (optional). */
+  name?: string;
+
+  /**
+   * Whether the health-pill portion of the tensor widget is to be
+   * included
+   *
+   * Defaults to `true`.
+   */
+  includeHealthPill?: boolean;
+
+  /** Defaults to `true`. */
+  includeMenu?: boolean;
+
+  /**
+   * How many decimal places to display the values in.
+   *
+   * The values of the tensor may be displaced in the decimal notation, or
+   * the engineering notation, depending automatically by the tensor-widget
+   * library based on the maximum absolute value of the elements of the tensor.
+   */
+  decimalPlaces?: number;
+
+  /** TODO(cais): Add support for custom tensor renderers. */
+}
+
+/**
+ * A TensorWidget that interactively visualizes a single tensor.
+ */
+export interface TensorWidget {
+  /**
+   * Renders the GUI of the tensor widget.
+   *
+   * This method should be called only once after the tensor widget is
+   * instantiated, or when the content of the underlying tensor has
+   * changed.
+   */
+  render: () => Promise<void>;
+
+  /**
+   * Scroll along the horizontal dimension.
+   *
+   * I.e., whichever dimension that's selected as the horizontal viewing
+   * dimension at the current time.
+   *
+   * `offset` will become the first element in the view, regardless of
```

Comment:
Sorry, this phrase doesn't parse in my head. `offset will become the first element in the view, regardless of...` 

What is offset supposed to be? Is it an index or is it the scroll position? If latter, what is the coordinate system (is negative towards left?)?

Reply:
It means index. I see that "index" would be a less confusing term. Done.